realtime/websocket/processors.py

```diff
@@ -0,0 +1,118 @@
+        """
+        self.audio_output_stream = audio_stream
+        self.message_stream = message_stream
+        self.video_stream = video_stream
+
+        audio_data = b""
+        while True:
+            try:
+                data = await self.ws.receive_json()
+                if data.get("type") == "message":
+                    await self.message_stream.put(data.get("data"))
+                elif data.get("type") == "audio":
+                    audio_bytes = base64.b64decode(data.get("data"))
+                    audio_data += audio_bytes
+
+                    if len(audio_data) < 2:
+                        continue
+                    if len(audio_data) % 2 != 0:
+                        # TODO: Get audio dtype from the frontend instead of hardcoding it
+                        array = np.frombuffer(audio_data[:-1], dtype=np.int16).reshape(1, -1)
+                        audio_data = audio_data[-1:]
+                    else:
+                        # TODO: Get audio dtype from the frontend instead of hardcoding it
+                        array = np.frombuffer(audio_data, dtype=np.int16).reshape(1, -1)
+                        audio_data = b""
+
+                    frame = av.AudioFrame.from_ndarray(array, format="s16", layout="mono")
+                    frame.sample_rate = 8000
+                    await self.audio_output_stream.put(frame)
+            except Exception as e:
+                logging.error("websocket: Exception", e)
+                raise asyncio.CancelledError()
+
+
+class WebsocketOutputStream:
+    """
+    Handles outgoing WebSocket messages by streaming audio and text data.
+
+    Attributes:
+        ws (WebSocket): The WebSocket connection.
+    """
+
+    def __init__(self, ws: WebSocket):
+        self.ws = ws
+
+    async def run(
+        self, audio_stream: AudioStream, message_stream: TextStream, video_stream: VideoStream, byte_stream: ByteStream
+    ):
+        """
+        Starts tasks to process and send audio and text streams.
+
+        Args:
+            audio_stream (AudioStream): The audio stream to send.
+            message_stream (TextStream): The text stream to send.
+            video_stream (VideoStream): The video stream to send.
+            byte_stream (ByteStream): The byte stream to send.
+        """
+        await asyncio.gather(self.task(byte_stream), self.task(message_stream))
+
+    async def task(self, input_stream):
+        """
+        Sends data from the input stream over the WebSocket.
+
+        Args:
+            input_stream (Stream): The stream from which to send data.
+        """
+        while True:
+            data = await input_stream.get()
+            if data is None:
+                json_data = {"type": "audio_end", "timestamp": time.time()}
+                await self.ws.send_json(json_data)
+            elif isinstance(data, bytes):
+                output_bytes_io = io.BytesIO()
+                in_memory_wav = wave.open(output_bytes_io, "wb")
```

Comment:
set to 16000 here

Reply:
??

Reply:
I meant it's 16kHz here as opposed to 8kHz in the WebsocketInputStream.

Reply:
added comment to update it in newer version